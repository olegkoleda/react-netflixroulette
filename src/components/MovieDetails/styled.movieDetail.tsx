import styled from "styled-components";

export const StyledImage = styled.img`
  width: 100%;
  height: 100%;
  object-fit: fill;
  max-height: 25rem;
`;

export const StyledRating = styled.span`
  width: 4rem;
  height: 4rem;
  border: thin solid #fff;
  color: green;
  font-size: 1.25rem;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  margin-left: 1rem;
`;

export const StyledTagLine = styled.span`
  font-size: 1rem;
  margin-top: 0.5rem;
`;

export const StyledAccent = styled.span`
  font-size: 1.5rem;
  color: ${(props) => props.theme.color.red};
  margin-right: 2rem;
`;
