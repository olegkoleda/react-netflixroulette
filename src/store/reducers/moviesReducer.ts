import {
  LOAD_MOVIES_ERROR,
  LOAD_MOVIES_STARTED,
  LOAD_MOVIES_SUCCESS,
  SET_ACTIVE_MOVIE,
  LoadMovieActionType,
  ISetActiveMovie,
} from "../types/loadMovies";
import {
  ADD_MOVIE_ERROR,
  ADD_MOVIE_STARTED,
  ADD_MOVIE_SUCCESS,
  ADD_MOVIE_FINISHED,
  AddMovieActionType,
} from "../types/addMovie";
import {
  DELETE_MOVIE_ERROR,
  DELETE_MOVIE_STARTED,
  DELETE_MOVIE_SUCCESS,
  DeleteMovieActionType,
} from "../types/deleteMovie";
import { IMovie } from "../../interfaces/IMovie";
import {
  UPDATE_MOVIE_ERROR,
  UPDATE_MOVIE_STARTED,
  UPDATE_MOVIE_SUCCESS,
  UpdateMovieActionType,
  UPDATE_MOVIE_FINISHED,
} from "../types/updateMovie";

export interface IMoviesState {
  list: IMovie[];
  loading: boolean;
  isMovieUpdated: boolean;
  error: Error | null;
  activeMovie: number | null;
  isMovieAdded: boolean;
}

const initialState: IMoviesState = {
  list: [],
  loading: false,
  error: null,
  activeMovie: null,
  isMovieUpdated: false,
  isMovieAdded: false,
};

export default function moviesReducer(
  state = initialState,
  action:
    | LoadMovieActionType
    | ISetActiveMovie
    | AddMovieActionType
    | DeleteMovieActionType
    | UpdateMovieActionType
) {
  switch (action.type) {
    case LOAD_MOVIES_STARTED:
    case ADD_MOVIE_STARTED:
    case DELETE_MOVIE_STARTED:
    case UPDATE_MOVIE_STARTED:
      return {
        ...state,
        loading: true,
      };
    case LOAD_MOVIES_SUCCESS:
      return {
        ...state,
        loading: false,
        list: action.payload,
      };
    case LOAD_MOVIES_ERROR:
    case ADD_MOVIE_ERROR:
    case DELETE_MOVIE_ERROR:
    case UPDATE_MOVIE_ERROR:
      return {
        ...state,
        error: action.error,
      };
    case SET_ACTIVE_MOVIE:
      return {
        ...state,
        activeMovie: action.payload,
      };
    case UPDATE_MOVIE_FINISHED:
      return {
        ...state,
        isMovieUpdated: false,
      };
    case ADD_MOVIE_FINISHED:
      return {
        ...state,
        isMovieAdded: false,
      };
    case ADD_MOVIE_SUCCESS:
      return {
        ...state,
        loading: false,
        isMovieAdded: true,
        list: [action.payload, ...state.list],
      };
    case DELETE_MOVIE_SUCCESS:
      return {
        ...state,
        loading: false,
        list: state.list.filter(({ id }) => id !== action.payload),
      };
    case UPDATE_MOVIE_SUCCESS:
      return {
        ...state,
        loading: false,
        isMovieUpdated: true,
        list: state.list.map((movie) =>
          movie.id === action.payload.id ? action.payload : movie
        ),
      };
    default:
      return state;
  }
}
