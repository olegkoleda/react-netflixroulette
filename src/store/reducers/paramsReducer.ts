import { FilterMovieActionType, SET_MOVIE_FILTER } from "../types/filterMovie";

export interface IParamsState {
  filter: string;
  sortBy: string;
}

const initialState: IParamsState = {
  filter: "",
  sortBy: "release_date",
};

export default function paramsReducer(
  state = initialState,
  action: FilterMovieActionType
) {
  switch (action.type) {
    case SET_MOVIE_FILTER:
      return {
        ...state,
        filter: action.payload,
      };

    default:
      return state;
  }
}
