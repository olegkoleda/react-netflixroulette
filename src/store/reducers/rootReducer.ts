import { combineReducers } from "redux";
import { IParamsState } from "./paramsReducer";
import { IMoviesState } from "./moviesReducer";
import movies from "./moviesReducer";
import params from "./paramsReducer";

export default combineReducers({
  movies,
  params,
});

export interface IAppState {
  movies: IMoviesState;
  params: IParamsState;
}
