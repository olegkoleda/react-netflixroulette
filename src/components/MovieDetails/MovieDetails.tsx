import React from "react";
import { Box, Composition } from "atomic-layout";
import Heading from "../Heading";
import {
  StyledImage,
  StyledRating,
  StyledTagLine,
  StyledAccent,
} from "./styled.movieDetail";
import { IMovie } from "../../interfaces/IMovie";
import { IAppState } from "../../store/reducers/rootReducer";
import { connect } from "react-redux";

interface IMovieDetailsProps {
  movie: IMovie;
}

const MovieDetails: React.FC<IMovieDetailsProps> = ({ movie }) => {
  const {
    title = "Default title",
    tagline = "Default tagline",
    vote_average = 5,
    release_date = "2019",
    poster_path = "#",
    overview = "default",
    runtime = 123,
  } = movie || {};

  const releaseDate = new Date(release_date).getFullYear();

  return (
    <Composition
      templateCols={"minmax(200px, 350px) 1fr"}
      gap={"1.5rem"}
      marginTop={"2rem"}
    >
      <div>
        <StyledImage src={poster_path} alt={""} />
      </div>
      <div>
        <Box flex alignItems={"center"}>
          <Heading as={"h2"}>{title}</Heading>
          <StyledRating>{vote_average}</StyledRating>
        </Box>
        <StyledTagLine>{tagline}</StyledTagLine>
        <Box flex marginVertical={"1.75rem"}>
          <StyledAccent>{releaseDate}</StyledAccent>
          <StyledAccent>{runtime} min</StyledAccent>
        </Box>
        <p>{overview}</p>
      </div>
    </Composition>
  );
};

const mapState = (state: IAppState) => ({
  movie: state.movies.activeMovieData,
});

export default connect(mapState)(MovieDetails);
