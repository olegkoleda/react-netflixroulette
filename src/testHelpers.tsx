import React from "react";
import renderer from "react-test-renderer";
import { theme } from "./styles/Theme";
import { ThemeProvider } from "styled-components";
import { BrowserRouter as Router } from "react-router-dom";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { initialAppState } from "./store/reducers/rootReducer";

const middleWares = [thunk];
const mockStore = configureMockStore(middleWares);
const store = mockStore(initialAppState);

export function renderWithTheme(component: React.ReactNode) {
  return renderer.create(
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  );
}

export const reduxWrapper = (Component: typeof React.Component | React.FC, data = {}) => (
  <Provider store={store}>
    <Router>
      <Component {...data}/>
    </Router>
  </Provider>
);
