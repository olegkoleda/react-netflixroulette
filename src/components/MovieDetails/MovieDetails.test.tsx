// @ts-nocheck
import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import MovieDetails from "./MovieDetails";
import { renderWithTheme } from "../../testHelpers";
import { initialAppState } from "../../store/reducers/rootReducer";

const middleWares = [thunk];
const mockStore = configureMockStore(middleWares);
const store = mockStore(initialAppState);

test("Cards list matches the Snapshot", () => {
  const component = renderWithTheme(
    <Provider store={store}>
      <Router>
        <MovieDetails />
      </Router>
    </Provider>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
