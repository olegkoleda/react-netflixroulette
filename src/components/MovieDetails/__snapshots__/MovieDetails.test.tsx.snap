// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards list matches the Snapshot 1`] = `
<div
  className="sc-AxirZ ldZlFH"
>
  <div>
    <img
      alt=""
      className="styledmovieDetail__StyledImage-sc-1d8hh6q-0 hbqvWI"
      src="#"
    />
  </div>
  <div>
    <div
      className="sc-AxjAm cCFrVs"
    >
      <h2
        className="Heading-sc-15b2y3d-0 fKqoJe"
      >
        Default title
      </h2>
      <span
        className="styledmovieDetail__StyledRating-sc-1d8hh6q-1 cExLRd"
      >
        5
      </span>
    </div>
    <span
      className="styledmovieDetail__StyledTagLine-sc-1d8hh6q-2 bgPwHO"
    >
      Default tagline
    </span>
    <div
      className="sc-AxjAm HJxfs"
    >
      <span
        className="styledmovieDetail__StyledAccent-sc-1d8hh6q-3 hNxqho"
      >
        2019
      </span>
      <span
        className="styledmovieDetail__StyledAccent-sc-1d8hh6q-3 hNxqho"
      >
        123
         min
      </span>
    </div>
    <p>
      default
    </p>
  </div>
</div>
`;
